sys/net/gnrc/network_layer/ipv6/nib/nib.c

```diff
@@ -0,0 +1,630 @@
+static void _handle_nbr_sol(kernel_pid_t iface, const ipv6_hdr_t *ipv6,
+                            const ndp_nbr_sol_t *nbr_sol, size_t icmpv6_len);
+static void _handle_nbr_adv(kernel_pid_t iface, const ipv6_hdr_t *ipv6,
+                            const ndp_nbr_adv_t *nbr_adv, size_t icmpv6_len);
+
+static bool _resolve_addr(const ipv6_addr_t *dst, kernel_pid_t iface,
+                          gnrc_pktsnip_t *pkt, gnrc_ipv6_nib_nc_t *nce,
+                          _nib_onl_entry_t *entry);
+
+static void _handle_snd_na(gnrc_pktsnip_t *pkt);
+
+/* interface flag checks */
+#if GNRC_IPV6_NIB_CONF_ROUTER
+static inline bool _is_rtr(const gnrc_ipv6_netif_t *netif)
+{
+    return (netif->flags & GNRC_IPV6_NETIF_FLAGS_ROUTER);
+}
+#endif
+/** @} */
+
+void gnrc_ipv6_nib_init(void)
+{
+    evtimer_event_t *tmp;
+
+    mutex_lock(&_nib_mutex);
+    for (evtimer_event_t *ptr = _nib_evtimer.events;
+         (ptr != NULL) && (tmp = (ptr->next), 1);
+         ptr = tmp) {
+        evtimer_del((evtimer_t *)(&_nib_evtimer), ptr);
+    }
+    _nib_init();
+    mutex_unlock(&_nib_mutex);
+}
+
+void gnrc_ipv6_nib_init_iface(kernel_pid_t iface)
+{
+    _nib_iface_t *nib_iface;
+
+    assert(iface > KERNEL_PID_UNDEF);
+    DEBUG("nib: Initialize interface %u\n", (unsigned)iface);
+    mutex_lock(&_nib_mutex);
+    nib_iface = _nib_iface_get(iface);
+    assert(nib_iface != NULL);
+    /* TODO:
+     * - set link-local address here for stateless address auto-configuration
+     *   and 6LN
+     * - join solicited nodes group of link-local address here for address
+     *   resolution here
+     * - join all router group of link-local address here on router node here
+     * - become an router advertising interface here on non-6LR here */
+
+    _init_iface_arsm(nib_iface);
+    nib_iface->rs_sent = 0;
+    nib_iface->na_sent = 0;
+#if GNRC_IPV6_NIB_CONF_ROUTER
+    nib_iface->last_ra = UINT32_MAX;
+    nib_iface->ra_sent = 0;
+#endif
+    mutex_unlock(&_nib_mutex);
+}
+
+int gnrc_ipv6_nib_get_next_hop_l2addr(const ipv6_addr_t *dst,
+                                      kernel_pid_t iface, gnrc_pktsnip_t *pkt,
+                                      gnrc_ipv6_nib_nc_t *nce)
+{
+    int res = 0;
+
+    mutex_lock(&_nib_mutex);
+    do {    /* XXX: hidden goto ;-) */
+        if (ipv6_addr_is_link_local(dst)) {
+            /* TODO: Prefix-based on-link determination */
+            if ((iface == KERNEL_PID_UNDEF) ||
+                !_resolve_addr(dst, iface, pkt, nce,
+                               _nib_onl_get(dst, iface))) {
+                res = -EHOSTUNREACH;
+                break;
+            }
+        }
+        else {
+            /* TODO: Off-link next hop determination */
+            res = -EHOSTUNREACH;
+        }
+    } while (0);
+    mutex_unlock(&_nib_mutex);
+    return res;
+}
+
+void gnrc_ipv6_nib_handle_pkt(kernel_pid_t iface, const ipv6_hdr_t *ipv6,
+                              const icmpv6_hdr_t *icmpv6, size_t icmpv6_len)
+{
```

Comment:
So many **TODOS** that could end up in a nicely documented fashion => doxygen (:

Reply:
I'm not sure if putting TODOs in doxygen is productive here...